File: 917 Reverse Only Letters.java
class Solution {
    public String reverseOnlyLetters(String s) {
          int i = 0;
    int j = s.length() - 1;
    
    char ch[] = s.toCharArray();
    
    while(i < j){
        
        if(!isAlphabet(ch[i])){
            i++;
            continue;
        }
        else if(!isAlphabet(ch[j])){
            j--;
            continue;
        }
        
        char cht = ch[i];
        ch[i] = ch[j];
        ch[j] = cht;
        
        i++;
        j--;
    }
 
    return new String(ch);

    }

    public boolean isAlphabet(char ch){

        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    }
}

File: 917 Reverse Only Letters.go
func reverseOnlyLetters(s string) string {
	ss := []byte(s)

	for i, j := 0, len(ss)-1; i<j; {
		if isAlpha(ss[i]) && isAlpha(ss[j]) {
			ss[i], ss[j] = ss[j], ss[i]
			i++
			j--
			continue
		}
		if !isAlpha(ss[i]) && !isAlpha(ss[j]) {
			i++
			j--
			continue
		}
		if isAlpha(ss[i]) {
			j--
			continue
		}
		i++
	}

	return string(ss)
}

func isAlpha(r byte) bool {
	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
}
